fdbserver/DataDistribution.actor.cpp

```diff
@@ -4251,26 +4263,42 @@ ACTOR Future<Void> perpetualStorageWiggleIterator(AsyncVar<bool>* stopSignal,
 	return Void();
 }
 
-// Watch the value change of `wigglingStorageServerPrefix`.
-// Return the watch future and the current value of `wigglingStorageServerPrefix`.
-ACTOR Future<std::pair<Future<Void>, Optional<Value>>> watchPerpetualStorageIDChange(DDTeamCollection* self) {
-	state ReadYourWritesTransaction tr(self->cx);
-	state Future<Void> watchFuture;
-	state Optional<Value> ret;
-	state const Key readKey = wigglingStorageServerPrefix.withSuffix(self->primary ? "primary"_sr : "remote"_sr);
+// read the current map of `perpetualStorageWiggleIDPrefix`, then restore wigglingId.
+ACTOR Future<Void> readStorageWiggleMap(DDTeamCollection* self) {
 
+	state const Key readKey = perpetualStorageWiggleIDPrefix.withSuffix(self->primary ? "primary/"_sr : "remote/"_sr);
+	state KeyBackedObjectMap<UID, StorageWiggleValue, decltype(IncludeVersion())> metadataMap(readKey,
+	                                                                                          IncludeVersion());
+	state Reference<ReadYourWritesTransaction> tr(new ReadYourWritesTransaction(self->cx));
+	state std::vector<std::pair<UID, StorageWiggleValue>> res;
+	// read the wiggling pairs
 	loop {
 		try {
-			tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
-			wait(store(ret, tr.get(readKey)));
-			watchFuture = tr.watch(readKey);
-			wait(tr.commit());
+			tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			wait(store(res, metadataMap.getRange(tr, UID(0, 0), Optional<UID>(), CLIENT_KNOBS->TOO_MANY)));
+			wait(tr->commit());
 			break;
 		} catch (Error& e) {
-			wait(tr.onError(e));
+			wait(tr->onError(e));
 		}
 	}
-	return std::make_pair(watchFuture, ret);
+	if (res.size() > 0) {
+		// SOMEDAY: support wiggle multiple SS at once
+		ASSERT(!self->wigglingId.present()); // only single process wiggle is allowed
```

Comment:
Is there a way to handle this gracefully?  If, for example, 7.2 allowed multiple items in this set, then downgrading to 7.1 would hit this `ASSERT` unless there were a patch for 7.1 to handle this without throwing an `internal_error`, and we wouldn't want to do a patch release just for that.

Reply:
What you said won't happen. This function will only run once when the storage_wiggler's status changes from `off` to `on` or when the DD restarted, the `wigglingId` is always an empty optional at this time no matter whether we support multiple SS. 
And because `wigglingId` is in-memory data, change `wigglingId` from a value to a set would be fine. I don't make an ASSERT on `res` (the persistent data).

Reply:
I see, sounds good!  I think I got confused by the `ASSERT` comment.